### ISSUE TITLE
Nested non-scalar types are not using its input type declarations

### ISSUE THREAD

**Issue by laurensk** [EntityGraphQL/EntityGraphQL#356]
For example, if we have a `User` and `Post` entity, where a user has posts.

I registered the entities for queries like this:

```cs
schema.AddType<User>("User");
schema.AddType<Post>("Post");

schema.UpdateType<User>(type =>
{
    type.AddField("id", x => x.Id, null);
    type.AddField("name", x => x.Name, null);
    type.AddField("posts", x => x.Posts, null);
});

schema.UpdateType<Post>(type =>
{
    type.AddField("id", x => x.Id, null);
    type.AddField("content", x => x.Content, null);
    type.AddField("userId", x => x.UserId, null);
    type.AddField("user", x => x.User, null);
});
```

Now, I want to create an input type which creates a user with some posts (in the same mutation). I have registered input types the following way:

```cs
var postInput = schema.AddInputType<Post>("PostInput");

userInput.AddField("content", x => x.Content, null);

var userInput = schema.AddInputType<User>("UserInput");

userInput.AddField("id", x => x.Id, null);
userInput.AddField("name", x => x.Name, null);
userInput.AddField("posts", x => x.Posts, null);
```

The problem now is that the type name on the generated schema for `posts` on the `UserInput` type is `Post` instead of `PostInput`, raising an error in GraphQL clients since the `Post` type is not an input type (it also does not have the properties of the input type, which I would want it to have).

Is there a way for the library to automatically lookup InputTypes of a type when used inside another InputType? Or is there a workaround to at least get this working in my case?

Thanks a lot!

**Comment by lukemurray** [comment]
This should be possible to fix as the rules in GraphQL are well defined for when it should be an InputType.

Currently it seems to look up by dotnet type only and gets the query type first as it was added first.

**Comment by lukemurray** [comment]
Thinking about this and playing with some code changes I think the better fix is throwing an error as #308 suggests. There is too much in the code base that relies on dotnet `Type` mapping to a single schema type.

You can easily create a `UserInput` & `PostInput` type to use that extends the query types.

```cs
public class PostInput : Post {}
public class UserInput : User {}

var postInput = schema.AddInputType<PostInput>("PostInput");
postInput.AddField("content", x => x.Content, null);

var userInput = schema.AddInputType<UserInput>("UserInput");
userInput.AddField("id", x => x.Id, null);
userInput.AddField("name", x => x.Name, null);
userInput.AddField("posts", x => x.Posts, null);
```

However there will still be a fix so that `userInput.AddField("posts", x => x.Posts, null)` finds the correct `PostInput` type. Which we can do.
